### ISSUE TITLE
Tuple deconstruct in foreach of an array returned by method from another assembly fails

### ISSUE THREAD

**Issue by NinoFloris** [dotnet/roslyn#41359]
Tuple deconstruct in foreach of an **array** (and only an array) returned by a **method** (and only a method) from another assembly fails.

Repro can be found here:
https://github.com/NinoFloris/roslyndeconstructbug
 
```cs
        static void Error()
        {
            // this here causes errors, see build output
            foreach (var (x, y) in Lib.Foo.Bar()) { }
        }

        static ValueTuple<Type, Type>[] BarSameAssembly()
            => new[] { (typeof(int), typeof(int)) };

        static void Success()
        {
            foreach (var (x, y) in BarSameAssembly()) { }
            foreach (var (x, y) in Lib.Foo.BarProp) { }
            foreach (var (x, y) in Lib.Foo.WorkingBar()) { }
        }
```

Build output:
```
Program.cs(9,36): error CS1061: 'ValueTuple<Type, Type>' does not contain a definition for 'Deconstruct' and no accessible extension method 'Deconstruct' accepting a first argument of type 'ValueTuple<Type, Type>' could be found (are you missing a using directive or an assembly reference?) [/Users/nfloris/Projects/Crowded/compilerbug/Console.csproj]
Program.cs(9,36): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'ValueTuple<Type, Type>', with 2 out parameters and a void return type. [/Users/nfloris/Projects/Crowded/compilerbug/Console.csproj]
Program.cs(9,27): error CS8130: Cannot infer the type of implicitly-typed deconstruction variable 'x'. [/Users/nfloris/Projects/Crowded/compilerbug/Console.csproj]
Program.cs(9,30): error CS8130: Cannot infer the type of implicitly-typed deconstruction variable 'y'. [/Users/nfloris/Projects/Crowded/compilerbug/Console.csproj]

The build failed. Fix the build errors and run again.
```

Pretty bizarre bug.

**Version Used**: dotnet sdk 3.1.101

**Comment by jcouv** [comment]
@NinoFloris I was not able to repro this. I cloned the repo you linked and ran `dotnet build`, which succeeded.
Can you confirm the problem still repro's for you? If so, can you add `#error version` in your C# source and let me know what version you get, so that we can compare.

```
D:\repos\roslyndeconstructbug>dotnet build
Microsoft (R) Build Engine version 16.7.0-preview-20360-03+188921e2f for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored D:\repos\roslyndeconstructbug\Lib\Lib.csproj (in 158 ms).
  Restored D:\repos\roslyndeconstructbug\Console.csproj (in 151 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  Lib -> D:\repos\roslyndeconstructbug\Lib\bin\Debug\netstandard2.0\Lib.dll
  Console -> D:\repos\roslyndeconstructbug\bin\Debug\netcoreapp3.1\Console.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.76
```

**Comment by NinoFloris** [comment]
Verified fixed, went back and all 3.1.1xx bands have it, 3.1.200 ships the fix.

**Comment by jcouv** [comment]
Thanks for confirming. I'll go ahead and close the issue.
